### PR TITLE
Update gather and scatter to not use Argument Encoder

### PR DESCRIPTION
**Pull request by jagrit06**
## Proposed changes

Virtual Apple silicon devices do not have support for argument encoders, which is a challenge for the CI pipeline. I just adjusted our gather and scatter kernels to not rely on  argument encoders while leaving the core function the same. No noticeable regression in performance was observed on  M3 Max 

Before:
```
====================
X (100, 64), Indices (1000000,)
MLX: 2.767ms
PyTorch: 10.588ms
====================
X (100, 1024), Indices (100000,)
MLX: 2.901ms
PyTorch: 16.767ms
====================
X (4, 1000000), Indices ()
MLX: 0.250ms
PyTorch: 0.622ms
```

After:
```
====================
X (100, 64), Indices (1000000,)
MLX: 2.214ms
PyTorch: 10.772ms
====================
X (100, 1024), Indices (100000,)
MLX: 2.893ms
PyTorch: 17.314ms
====================
X (4, 1000000), Indices ()
MLX: 0.211ms
PyTorch: 0.234ms
```

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Comment by jagrit06** [comment]
Fixes issue #31

**Review by awni — COMMENTED** [review]
LGTM once it passes the tests. Getting rid of those argument buffers is a pretty nice simplification!

**Comment by jagrit06** [comment]
I don’t think we’d call it a bug - you can check the MTLDevice object to see if it has support for argument encoders and that call comes out false for virtual devices 

I suppose it has to with packing pointers in a buffer and making that redirection available requires something more (there are also some hardware mechanics involved for some versions / things you are able do with argument encoders)

File: mlx/backend/metal/indexing.cpp [hunk 1]
```diff
@@ -63,91 +64,47 @@ void Gather::eval_gpu(const std::vector<array>& inputs, array& out) {
...
-  }
+        inputs[i + 1].shape().end());
 
-  // Allocate the argument buffer
-  auto arg_buf = allocator::malloc_or_wait(arg_enc->encodedLength());
-
-  // Register data with the encoder
-  arg_enc->setArgumentBuffer(static_cast<MTL::Buffer*>(arg_buf.ptr()), 0);
-  for (int i = 0; i < nidx; ++i) {
-    set_array_buffer(compute_encoder, arg_enc, inputs[i + 1], i);
```
Comment (awni) [inline mlx/backend/metal/indexing.cpp:114]: Nit: let's remove this function overload since we should never use it

Comment (awni) [reply mlx/backend/metal/indexing.cpp:114]: I think it's also worth removing the argument encoder support from device.cpp, wdyt?

Comment (awni) [reply mlx/backend/metal/indexing.cpp:114]: I just love deleting code 😛

Comment (jagrit06) [reply mlx/backend/metal/indexing.cpp:114]: Agree, it will just be confusing

**Pull request merged**